libflux/src/semantic/nodes.rs

```diff
@@ -573,8 +603,72 @@ pub struct BinaryExpr {
 }
 
 impl BinaryExpr {
-    fn infer(&self, env: Environment, f: &mut Fresher) -> Result {
-        unimplemented!();
+    fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
+        // Compute the left and right constraints.
+        // Do this first so that we can return an error if one occurs.
+        let (env, lcons) = self.left.infer(env, f)?;
+        let (env, rcons) = self.right.infer(env, f)?;
+
+        let mut cons = Vec::new();
+        if let Some(kind) = match self.operator {
+            ast::Operator::AdditionOperator => Some(Addable),
+            ast::Operator::SubtractionOperator => Some(Subtractable),
+            ast::Operator::MultiplicationOperator
+            | ast::Operator::DivisionOperator => Some(Divisible),
+            _ => None,
+        } {
+            cons.push(Constraint::Kind(self.typ.clone(), kind));
+        }
+
+        cons.append(&mut Vec::from(match self.operator {
+            // The following operators require both sides to be equal.
+            ast::Operator::AdditionOperator
+            | ast::Operator::SubtractionOperator
+            | ast::Operator::MultiplicationOperator
+            | ast::Operator::DivisionOperator
+            | ast::Operator::PowerOperator
+            | ast::Operator::ModuloOperator => Constraints::from(vec![
+                Constraint::Equal(self.left.type_of().clone(), self.typ.clone()),
+                Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
+            ]),
+            // The following require the type to be a boolean.
+            ast::Operator::GreaterThanEqualOperator
+            | ast::Operator::LessThanEqualOperator
+            | ast::Operator::GreaterThanOperator
+            | ast::Operator::LessThanOperator
+            | ast::Operator::NotEqualOperator
+            | ast::Operator::EqualOperator => {
+                let kind = match self.operator {
+                    ast::Operator::EqualOperator
+                    | ast::Operator::NotEqualOperator => Equatable,
+                    _ => Comparable,
+                };
+                Constraints::from(vec![
+                    Constraint::Kind(self.left.type_of().clone(), kind),
+                    Constraint::Kind(self.right.type_of().clone(), kind),
+                    Constraint::Equal(self.typ.clone(), MonoType::Bool),
```

Comment:
You're missing a constraint here. Both sides should be constrained to be equal.

Reply:
This is not one of the constraints we have in the Go code. It requires polymorphic literals for this to be the case.